src/js.rs

```diff
@@ -0,0 +1,761 @@
+    let elements = elements.into_iter().map(|e| 
+        maybe_block_expr(e, &env)
+    );
+    let content = concat(Itertools::intersperse(elements, break_(",", ", ")));
+    let content = if let Some(final_tail) = final_tail {
+        content.append(Document::String(", ...".to_string()).append(maybe_block_expr(final_tail, &env)))
+    } else {
+        content
+    };
+    content.surround("[", "]")
+}
+
+fn collect_cons<'a>(tail: &'a TypedExpr, elements: &mut Vec<&'a TypedExpr>) -> Option<&'a TypedExpr> {
+    match tail {
+        TypedExpr::ListNil { .. } => None,
+        TypedExpr::ListCons { head, tail, .. } => {
+            elements.push(head);
+            collect_cons(tail, elements)
+        }
+        // TODO is it possible to write improper lists in Gleam
+        other => Some(other)
+    }
+}
+
+fn tuple<'a>(elems: impl Iterator<Item = Document<'a>>) -> Document<'a> {
+    concat(Itertools::intersperse(elems, break_(",", ", ")))
+        .nest_current()
+        .surround("[", "]")
+        .group()
+}
+
+fn tuple_index<'a>(tuple: &'a TypedExpr, index: u64) -> Document<'a> {
+    let env = Env{ return_last: &false, semicolon: &false};
+    expr(tuple, &env)
+        .append(index.to_doc().surround("[", "]"))
+}
+
+fn call<'a>(fun: &'a TypedExpr, args: &'a [CallArg<TypedExpr>]) -> Document<'a> {
+    let args = args.into_iter().map(|arg| 
+        maybe_block_expr(&arg.value, &Env{ return_last: &false, semicolon: &false})
+    );
+    match fun {
+        TypedExpr::Var {
+            constructor:
+                ValueConstructor {
+                    variant: ValueConstructorVariant::ModuleFn { module, name, .. },
+                    ..
+                },
+            ..
+        } => {
+            // TODO self module or not
+            Document::String(name.to_string())
+            .append(concat(Itertools::intersperse(args, break_(",", ", "))).surround("(", ")"))
+
+        }
+        TypedExpr::Var {
+            constructor:
+                ValueConstructor {
+                    variant: ValueConstructorVariant::LocalVariable,
+                    ..
+                },
+            name,
+            ..
+        } => {
+            // TODO self module or not
+            Document::String(name.to_string())
+            .append(concat(Itertools::intersperse(args, break_(",", ", "))).surround("(", ")"))
+        }
+        TypedExpr::Var {
+            constructor:
+                ValueConstructor {
+                    variant: ValueConstructorVariant::Record { name, field_map, .. },
+                    ..
+                },
+            ..
+        } => {
+            // println!("args: {:?}", args);
+            let fields = match field_map {
+                Some(FieldMap{fields, ..}) => {
+                    // fields.into_iter()
+                    // TODO Can I guarantee the order here?
+                    
+                    let t = fields
+                    .iter();
```

Comment:
How can I get a vector of map keys ordered by the value they contain?

The `fields` value here looks like `{"name": 0, "age": 1}`
I want the ordered names so I can zip with the ordered args and generate the code `{type: "Cat", name: "example", age: 1}`

Reply:
```rs
fields.iter().sorted_by(|(_, &v)| v).map(|(name, _)| name)
```

is roughly what you need

Reply:
Well it's sounds obvious when you put it like that.  Thank you.

Reply:
Just in case it's not obvious, not all arguments are guaranteed to have labels so positional and labelled arguments will need to be merged.